src/TorchSharp/NN/Optimizer.cs

```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
+            /// </summary>
+            /// <param name="parameters">Parameters to optimize. This optimizer requires the <b>named</b> parameters collection.</param>
+            /// <param name="learningRate">Learning rate</param>
+            /// <param name="momentum">Momentum factor (default: 0)</param>
+            /// <param name="dampening">Dampening for momentum (default: 0)</param>
+            /// <param name="weight_decay">Weight decay (L2 penalty) (default: 0)</param>
+            /// <param name="nesterov">Enables Nesterov momentum (default: False)</param>
+            /// <param name="maximize"></param>
+            /// <returns></returns>
+            public static Modules.SGD SGD(IEnumerable<SGD.ParamGroup> parameters, double learningRate, double momentum = 0.0, double dampening = 0, double weight_decay = 0, bool nesterov = false, bool maximize = false)
+            {
+                return new Modules.SGD(parameters, learningRate, momentum, dampening, weight_decay, nesterov, maximize);
+            }
+        }
+    }
+
+    namespace Modules
+    {
+        using static torch.optim;
+
+        // Most optimizers are implemented in native code, but a few of them are directly implemented in
+        // managed code.
+
+        /// <summary>
+        /// Base class to help with a couple of the things that managed-code implementations need.
+        /// </summary>
+
+        public class OptimizerHelper : Optimizer
+        {
+            public OptimizerHelper() : base(IntPtr.Zero)
             {
-                LearningRate = learningRate;
-                InitialLearningRate = learningRate;
-                _parameters = named_parameters;
             }
 
             public override void zero_grad()
             {
-                foreach (var (_, p) in _parameters) {
+                foreach (var g in _parameter_groups) {
+
+                    foreach (var p in g.Parameters) {
+
+                        using var grad = p.grad();
+
+                        if (grad is null) continue;
+
+                        grad.zero_().Dispose();
+                    }
+                }
+            }
+
+            protected Tensor _step<T>(Action<T> body, Func<Tensor> loss_closure = null) where T:ParamGroup
+            {
+                Tensor loss = null;
+
+                if (loss_closure != null) {
+                    using (var _ = torch.enable_grad())
+                        loss = loss_closure();
+                }
+
+                using (var _ = torch.no_grad()) {
+
+                    using (var d = torch.NewDisposeScope()) {
+
+                        foreach (var group in _parameter_groups) {
 
-                    using var grad = p.grad();
+                            body(group as T);
 
-                    if (grad is null) continue;
+                        }
 
-                    grad.zero_().Dispose();
+                        d.DisposeEverything();
```

Comment:
Why do you have to call this manually? I thought you did not need to, it would be called by the implicit `d.Dispose()`?